botorch/acquisition/proximal.py

```diff
@@ -0,0 +1,128 @@
+
+r"""
+A wrapper around AquisitionFunctions to fix certain features for optimization.
+This is useful e.g. for performing contextual optimization.
+"""
+
+from __future__ import annotations
+
+import torch
+from botorch.acquisition.analytic import AnalyticAcquisitionFunction
+from botorch.utils import t_batch_mode_transform
+from torch import Tensor
+from torch.nn import Module
+
+
+class ProximalAcquisitionFunction(AnalyticAcquisitionFunction):
+    """A wrapper around AquisitionFunctions to add proximal weighting of the acquisition function. Acquisition
+    function is weighted via a squared exponential centered at the last training point, with varying lengthscales
+    corresponding to 'proximal_weights'. Can only be used with single batch analytical acquisition functions.
+
+    Example:
+        >>> model = SingleTaskGP(train_X, train_Y)  # d = 2
+        >>> EI = ExpectedImprovement(model, best_f=0.0)
+        >>> proximal_weights = torch.ones(d)
+        >>> EI_proximal = ProximalAcquisitionFunction(EI, proximal_weights)
+        >>> eip = EI_proximal(test_X)  # d' = 3
+    """
+
+    def __init__(
+        self,
+        acq_function: AnalyticAcquisitionFunction,
+        proximal_weights: Tensor,
+    ) -> None:
+        r"""Derived Acquisition Function by weighted by proximity to recently observed point.
+
+        Args:
+            acq_function: The base acquisition function, operating on input
+                tensors of feature dimension `d`.
+            proximal_weights: Tensor used to bias locality along each axis, should be in the shape of ('d',).
+
+        """
+        Module.__init__(self)
+        self.acq_func = acq_function
+
+        self.register_buffer("proximal_weights", proximal_weights)
+
+        # check to make sure that weights match the training data shape
+        assert (
+            self.proximal_weights.shape[0]
+            == self.acq_func.model.train_inputs[0][-1].shape[-1]
+        )
+
+    @t_batch_mode_transform(expected_q=1, assert_output_shape=False)
+    def forward(self, X: Tensor):
+        r"""Evaluate base acquisition function with proximal weighting.
+
+        Args:
+            X: Input tensor of feature dimension `d` .
+
+        Returns:
+            Base acquisition function evaluated on tensor `X` multiplied by a squared exponenital centered at the last
+            training point and is weighted according to the weighting of each dimension.
+        """
+        d = self.proximal_weights.shape[0]
+        last_X = self.acq_func.model.train_inputs[0][-1].reshape(1, 1, -1)
+        _unbroadcasted_scale_tril = torch.diag(
+            torch.sqrt(self.proximal_weights)
+        ).reshape(1, d, d)
+
+        diff = X - last_X
+        M = _batch_mahalanobis(_unbroadcasted_scale_tril, diff)
+        proximal_acq_weight = torch.exp(-0.5 * M)
+        return self.acq_func(X) * proximal_acq_weight.flatten()
+
+
+def _batch_mahalanobis(bL, bx):
```

Comment:
This looks suspiciously like `torch.distributions.multivariate_normal._batch_mahalanobis` - maybe just import that?

Reply:
Actually, since `_unbroadcasted_scale_tril` is always diagonal by construction, this is kind of boiling the ocean - the whole thing above just be equivalent to `torch.sum(diff ** 2 / self.proximal_weights)` and you can get rid of this function altogether.

Reply:
I ended up using the torch.distributions version of _batch_mahalanobis, it was easier to implement the calculation in batch model than what you stated. Initially I was being over-conservative with my imports